packages/devtools_app/lib/src/ui/hover.dart

```diff
@@ -358,74 +368,139 @@ class _HoverCardTooltipState extends State<HoverCardTooltip> {
     _removeTimer = null;
 
     if (!widget.enabled()) return;
-    _showTimer = Timer(HoverCardTooltip._hoverDelay, () async {
-      HoverCardData? hoverCardData;
+    final asyncGenerateHoverCardData = widget.asyncGenerateHoverCardData;
+    final generateHoverCardData = widget.generateHoverCardData;
+    final asyncTimeout = widget.asyncTimeout;
 
-      if (widget.asyncGenerateHoverCardData != null) {
-        assert(widget.generateHoverCardData == null);
-        // The data on the card is fetched asynchronously, so show a spinner
-        // while we wait for it.
-        final spinnerHoverCard = HoverCard.fromHoverEvent(
-          context: context,
-          contents: const CenteredCircularProgressIndicator(),
-          width: HoverCardTooltip.defaultHoverWidth,
+    _showTimer = Timer(HoverCardTooltip._hoverDelay, () async {
+      if (asyncGenerateHoverCardData != null) {
+        assert(generateHoverCardData == null);
+        _showAsyncHoverCard(
+          asyncGenerateHoverCardData: asyncGenerateHoverCardData,
           event: event,
-          hoverCardController: _hoverCardController,
+          asyncTimeout: asyncTimeout,
         );
-
-        _setHoverCard(
-          spinnerHoverCard,
+      } else {
+        _setHoverCardFromData(
+          generateHoverCardData!(event),
+          context: context,
+          event: event,
         );
+      }
+    });
+  }
 
-        // The spinner is showing, we can now generate the HoverCardData
-        hoverCardData = await widget.asyncGenerateHoverCardData!(
+  void _showAsyncHoverCard({
+    required AsyncGenerateHoverCardDataFunc asyncGenerateHoverCardData,
+    required PointerHoverEvent event,
+    int? asyncTimeout,
+  }) async {
+    HoverCard? spinnerHoverCard;
+    final hoverCardDataFuture = asyncGenerateHoverCardData(
+      event: event,
+      isHoverStale: () =>
+          spinnerHoverCard != null &&
+          !_hoverCardController.isHoverCardStillActive(spinnerHoverCard),
+    );
+    // If we have set the async hover card to show up only after a timeout,
+    // then race the timeout against generating the hover card data. If
+    // generating the data completes first, immediately show the hover card
+    // (or return early if there is no data).
+    if (asyncTimeout != null) {
+      final hoverCardDataCompleter =
```

Comment:
Elliott this is such a great use of completers and future.any! Such a slick way of approaching this problem!